[Cocoa] Teach CDMInstanceSessionFairPlayStreamingAVFObjC to create a built-in content key group behind an off-by-default setting

aestes opened the pull request:
#### 2a187984c01077d29f3963a2bb935cd6a8b70420
<pre>
[Cocoa] Teach CDMInstanceSessionFairPlayStreamingAVFObjC to create a built-in content key group behind an off-by-default setting
<a href="https://bugs.webkit.org/show_bug.cgi?id=261522">https://bugs.webkit.org/show_bug.cgi?id=261522</a>
rdar://115440398

Reviewed by Jer Noble.

Added a new, unstable (i.e. off-by-default) web preference named
BuiltInCDMKeyGroupingStrategyEnabled. When this preference is enabled,
CDMInstanceSessionFairPlayStreamingAVFObjC will create a WebContentKeyGroup instead of an
AVContentKeyReportGroup for managing CDM key reporting groups. WebContentKeyGroup is currently just
a stub but will gain a full implementation in follow-on commits.

No new tests; no change in behavior.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/encryptedmedia/MediaKeySession.cpp:
(WebCore::MediaKeySession::generateRequest):
(WebCore::MediaKeySession::keyGroupingStrategy const):
* Source/WebCore/Modules/encryptedmedia/MediaKeySession.h:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/graphics/cocoa/WebAVContentKeyReportGroupExtras.h: Copied from Source/WebCore/PAL/pal/graphics/cocoa/WebContentKeyGrouping.h.
* Source/WebCore/PAL/pal/graphics/cocoa/WebContentKeyGrouping.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/encryptedmedia/CDMInstanceSession.h:
* Source/WebCore/platform/encryptedmedia/CDMKeyGroupingStrategy.h: Copied from Source/WebCore/PAL/pal/graphics/cocoa/WebContentKeyGrouping.h.
* Source/WebCore/platform/encryptedmedia/CDMMediaKeyGroupingStrategy.h: Copied from Source/WebCore/PAL/pal/graphics/cocoa/WebContentKeyGrouping.h.
* Source/WebCore/platform/encryptedmedia/clearkey/CDMClearKey.cpp:
(WebCore::CDMInstanceSessionClearKey::requestLicense):
* Source/WebCore/platform/encryptedmedia/clearkey/CDMClearKey.h:
* Source/WebCore/platform/graphics/avfoundation/objc/CDMInstanceFairPlayStreamingAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/CDMInstanceFairPlayStreamingAVFObjC.mm:
(WebCore::CDMInstanceSessionFairPlayStreamingAVFObjC::requestLicense):
(WebCore::CDMInstanceSessionFairPlayStreamingAVFObjC::didProvideRequest):
(WebCore::CDMInstanceSessionFairPlayStreamingAVFObjC::ensureSessionOrGroup):
* Source/WebCore/platform/graphics/avfoundation/objc/ContentKeyGroupFactoryAVFObjC.h: Copied from Source/WebCore/PAL/pal/graphics/cocoa/WebContentKeyGrouping.h.
* Source/WebCore/platform/graphics/avfoundation/objc/ContentKeyGroupFactoryAVFObjC.mm: Copied from Source/WebCore/PAL/pal/graphics/cocoa/WebContentKeyGrouping.h.
* Source/WebCore/platform/graphics/avfoundation/objc/WebContentKeyGroup.h: Copied from Source/WebCore/PAL/pal/graphics/cocoa/WebContentKeyGrouping.h.
* Source/WebCore/platform/graphics/avfoundation/objc/WebContentKeyGroup.mm: Copied from Source/WebCore/PAL/pal/graphics/cocoa/WebContentKeyGrouping.h.
* Source/WebCore/testing/MockCDMFactory.cpp:
(WebCore::MockCDMInstanceSession::requestLicense):
* Source/WebCore/testing/MockCDMFactory.h:
* Source/WebKit/GPUProcess/media/RemoteCDMInstanceSessionProxy.cpp:
(WebKit::RemoteCDMInstanceSessionProxy::requestLicense):
* Source/WebKit/GPUProcess/media/RemoteCDMInstanceSessionProxy.h:
* Source/WebKit/GPUProcess/media/RemoteCDMInstanceSessionProxy.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/media/RemoteCDMInstanceSession.cpp:
(WebKit::RemoteCDMInstanceSession::requestLicense):
* Source/WebKit/WebProcess/GPU/media/RemoteCDMInstanceSession.h:

Canonical link: <a href="https://commits.webkit.org/267979@main">https://commits.webkit.org/267979@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a7f9a3ecf6c2bd1a7690582f4cf83b1a80c0e3bf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18215 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18549 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19126 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20054 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17050 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18410 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21846 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18705 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18994 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18436 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18647 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15857 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20937 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15871 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16610 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23114 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/15785 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16890 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16781 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20998 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/17496 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17349 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14712 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/21572 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16443 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5284 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4346 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20807 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/22805 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17196 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5131 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/platform/encryptedmedia/CDMInstanceSession.h:
```diff
@@ -79,7 +82,7 @@ class CDMInstanceSession : public RefCounted<CDMInstanceSession> {
     };
 
     using LicenseCallback = CompletionHandler<void(Ref<SharedBuffer>&& message, const String& sessionId, bool needsIndividualization, SuccessValue succeeded)>;
-    virtual void requestLicense(LicenseType, const AtomString& initDataType, Ref<SharedBuffer>&& initData, LicenseCallback&&) = 0;
+    virtual void requestLicense(LicenseType, KeyGroupingStrategy, const AtomString& initDataType, Ref<SharedBuffer>&& initData, LicenseCallback&&) = 0;
```
Looks like this change is causing problems for other ports. Maybe have two `virtual` methods, and have the new one just assume `CDMKeyGroupingStrategy::Platform` by default?

aestes committed:
[Cocoa] Teach CDMInstanceSessionFairPlayStreamingAVFObjC to create a built-in content key group behind an off-by-default setting
https://bugs.webkit.org/show_bug.cgi?id=261522
rdar://115440398

Reviewed by Jer Noble.

Added a new, unstable (i.e. off-by-default) web preference named
BuiltInCDMKeyGroupingStrategyEnabled. When this preference is enabled,
CDMInstanceSessionFairPlayStreamingAVFObjC will create a WebContentKeyGroup instead of an
AVContentKeyReportGroup for managing CDM key reporting groups. WebContentKeyGroup is currently just
a stub but will gain a full implementation in follow-on commits.

No new tests; no change in behavior.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/encryptedmedia/MediaKeySession.cpp:
(WebCore::MediaKeySession::generateRequest):
(WebCore::MediaKeySession::keyGroupingStrategy const):
* Source/WebCore/Modules/encryptedmedia/MediaKeySession.h:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/graphics/cocoa/WebAVContentKeyReportGroupExtras.h: Copied from Source/WebCore/PAL/pal/graphics/cocoa/WebContentKeyGrouping.h.
* Source/WebCore/PAL/pal/graphics/cocoa/WebContentKeyGrouping.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/encryptedmedia/CDMInstanceSession.h:
* Source/WebCore/platform/encryptedmedia/CDMKeyGroupingStrategy.h: Copied from Source/WebCore/PAL/pal/graphics/cocoa/WebContentKeyGrouping.h.
* Source/WebCore/platform/encryptedmedia/CDMMediaKeyGroupingStrategy.h: Copied from Source/WebCore/PAL/pal/graphics/cocoa/WebContentKeyGrouping.h.
* Source/WebCore/platform/encryptedmedia/clearkey/CDMClearKey.cpp:
(WebCore::CDMInstanceSessionClearKey::requestLicense):
* Source/WebCore/platform/encryptedmedia/clearkey/CDMClearKey.h:
* Source/WebCore/platform/graphics/avfoundation/objc/CDMInstanceFairPlayStreamingAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/CDMInstanceFairPlayStreamingAVFObjC.mm:
(WebCore::CDMInstanceSessionFairPlayStreamingAVFObjC::requestLicense):
(WebCore::CDMInstanceSessionFairPlayStreamingAVFObjC::didProvideRequest):
(WebCore::CDMInstanceSessionFairPlayStreamingAVFObjC::ensureSessionOrGroup):
* Source/WebCore/platform/graphics/avfoundation/objc/ContentKeyGroupFactoryAVFObjC.h: Copied from Source/WebCore/PAL/pal/graphics/cocoa/WebContentKeyGrouping.h.
* Source/WebCore/platform/graphics/avfoundation/objc/ContentKeyGroupFactoryAVFObjC.mm: Copied from Source/WebCore/PAL/pal/graphics/cocoa/WebContentKeyGrouping.h.
* Source/WebCore/platform/graphics/avfoundation/objc/WebContentKeyGroup.h: Copied from Source/WebCore/PAL/pal/graphics/cocoa/WebContentKeyGrouping.h.
* Source/WebCore/platform/graphics/avfoundation/objc/WebContentKeyGroup.mm: Copied from Source/WebCore/PAL/pal/graphics/cocoa/WebContentKeyGrouping.h.
* Source/WebCore/testing/MockCDMFactory.cpp:
(WebCore::MockCDMInstanceSession::requestLicense):
* Source/WebCore/testing/MockCDMFactory.h:
* Source/WebKit/GPUProcess/media/RemoteCDMInstanceSessionProxy.cpp:
(WebKit::RemoteCDMInstanceSessionProxy::requestLicense):
* Source/WebKit/GPUProcess/media/RemoteCDMInstanceSessionProxy.h:
* Source/WebKit/GPUProcess/media/RemoteCDMInstanceSessionProxy.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/media/RemoteCDMInstanceSession.cpp:
(WebKit::RemoteCDMInstanceSession::requestLicense):
* Source/WebKit/WebProcess/GPU/media/RemoteCDMInstanceSession.h:

Canonical link: https://commits.webkit.org/267979@main
